### PR TITLE
feat: add compatibility with individual date extensions for submissions

### PR DESCRIPTION
**Pull request by Ian2012**
**TL;DR -** Adds compatibility with individual date extensions behind a waffle flag

Partially fixes: https://github.com/openedx/edx-ora2/issues/1959

**What changed?**

- If the `due` date of the submission is greater than the `submission_due` date, `due` is used for the `submission_range`.
- If the `due` date of the assessment step is greater than the `assessment_due` date, `due` is used for the `assesment_step` range.

**Developer Checklist**

- [ ] Reviewed the [release process](https://github.com/openedx/edx-ora2/blob/master/.github/release_process.md)
- [ ] Translations and JS/SASS compiled
- [ ] Bumped version number in [openassessment/\_\_init\_\_.py](https://github.com/openedx/edx-ora2/blob/master/openassessment/__init__.py#L4) and [package.json](https://github.com/openedx/edx-ora2/blob/master/package.json#L3)

**Testing Instructions**

Add a waffle flag for everyone: `openresponseassessment.due_date_extension`

1. Go to the CMS
2. Add a default ORA component with any assessment steps.
3. Set a due date for the Section to yesterday.
4. Verify user can't answer the ORA.
5. Go to the instructor tab, extensions, and grant an individual due date for that section.
6. Verify user can answer the problem.

Before this change, users were unable to answer:

image

After setting a due date for yesterday:

image

After granting an individual date extension:

image

The second commit allows us to apply the individual due date extension for every step in the workflow:

https://github.com/openedx/edx-ora2/pull/2026/commits/c005190f6f2cf137183ae12f9a566c64d7d208c6

image

image

**Things to consider**

Unit tests will be added once we receive our first product review and concerns are resolved

**Reviewer Checklist**

Collectively, these should be completed by reviewers of this PR:

- [ ] I've done a visual code review
- [ ] I've tested the new functionality

FYI: @openedx/content-aurora

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @Ian2012!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

**Comment by Ian2012** [comment]
@e0d do you think this is a good start?

I was planning on also granting extensions for other steps, but I'm unaware if this is required for ORA individual due date extensions. In that case, could you provide a description for what's the expected workflow?

**Comment by Ian2012** [comment]
The second commit allows us to apply the individual due date extension for every step in the workflow:

https://github.com/openedx/edx-ora2/pull/2026/commits/c005190f6f2cf137183ae12f9a566c64d7d208c6

File: openassessment/xblock/openassessmentblock.py [hunk 1]
```diff
@@ -1106,11 +1106,23 @@ def is_closed(self, step=None, course_staff=None):
...
+                if asmnt_due is None or parse_date_value(self.due, self._) < parse_date_value(asmnt_due, self._):
+                    assessment_ranges.append((asmnt.get('start'), asmnt_due))
+                else:
+                    assessment_ranges.append((asmnt.get('start'), self.due))                
+        else:
+            submission_range = (self.submission_start, self.submission_due)
+            assessment_ranges = [
+                (asmnt.get('start'), asmnt.get('due'))
+                for asmnt in self.valid_assessments
+            ]
```
Comment (mariajgrimaldi) [inline openassessment/xblock/openassessmentblock.py:1125]: We should be able to turn this on and off. So can we put this behind a feature flag?

File: openassessment/xblock/openassessmentblock.py [hunk 2]
```diff
@@ -1106,11 +1106,23 @@ def is_closed(self, step=None, course_staff=None):
...
 
         """
-        submission_range = (self.submission_start, self.submission_due)
-        assessment_ranges = [
-            (asmnt.get('start'), asmnt.get('due'))
-            for asmnt in self.valid_assessments
-        ]
+        if self.due is not None \
+           and self.submission_due is not None \
+           and parse_date_value(self.due, self._) > parse_date_value(self.submission_due, self._):
```
Comment (mariajgrimaldi) [inline openassessment/xblock/openassessmentblock.py:1111]: This is hard to read. Can we refactor it?
```suggestion
        due_date_greater = parse_date_value(self.due, self._) > parse_date_value(self.submission_due, self._):
        if self.due and self.submission_due and due_date_greater
```

File: openassessment/xblock/openassessmentblock.py (hunk 1 above)
Comment (mariajgrimaldi) [inline openassessment/xblock/openassessmentblock.py:1125]: This block is hard to read because of how nested it is, can we try refactoring it a bit?

**Review by mariajgrimaldi — CHANGES_REQUESTED** [review]

File: openassessment/xblock/openassessmentblock.py [hunk 3]
```diff
@@ -1106,11 +1106,23 @@ def is_closed(self, step=None, course_staff=None):
...
+            submission_range = (self.submission_start, self.due)
+            assessment_ranges = []
+            for asmnt in self.valid_assessments:
```
Comment (mariajgrimaldi) [inline openassessment/xblock/openassessmentblock.py:1114]:
```suggestion
            for assessment in self.valid_assessments:
```

Comment (mariajgrimaldi) [reply openassessment/xblock/openassessmentblock.py:1114]: Consider this comment for the other occurrences

File: openassessment/xblock/openassessmentblock.py (hunk 1 above)
Comment (mariajgrimaldi) [reply openassessment/xblock/openassessmentblock.py:1125]: I think we can extract that piece of code into its own method and call it if and only if the feature is on.

**Comment by mariajgrimaldi** [comment]
I'm missing an explanation of why this is the change we need. And also, we need green checks!

File: openassessment/xblock/openassessmentblock.py [hunk 4]
```diff
@@ -1066,6 +1066,67 @@ def render_error(self, error_msg):
         template = get_template('openassessmentblock/oa_error.html')
         return Response(template.render(context), content_type='application/html', charset='UTF-8')
 
+    def can_use_date_extensions(self):
```
Comment (mariajgrimaldi) [inline openassessment/xblock/openassessmentblock.py:1069]: Missing a docstring here

**Review by mariajgrimaldi — COMMENTED** [review]
I left a few more comments!

**Comment by Ian2012** [comment]
This PR is our first proposal to add this functionality, but we have only an edge case in which we would like your opinion:

If the Subsection due date is configured to any date after the `submission_due`, the subsection due date is used instead. But we don't think this should be the case.

One solution we have in mind is to query for a [UserDate](https://github.com/openedx/edx-when/blob/b2c168dec435cc6a3bed93016688a14f8a113337/edx_when/models.py#L113-L118)  (which stores individual date extensions) object related to the subsection and to the student.

What do you think? @Colin-Fredericks @Daniel-hershel

**Comment by itsjeyd** [comment]
@Ian2012 Thanks for these changes!

Hi @jmakowski1123, could you please create a feature ticket for this PR? Or will we not need one since there is an existing issue (https://github.com/openedx/edx-ora2/issues/1959) that is being (partially) fixed here?

CC @e0d @mphilbrick211

**Comment by Daniel-hershel** [comment]
@Ian2012 I just wanted to give you a heads up that we are planning to launch a very similar feature in the next week or so. I am going to meet with the lead engineer working on that project later today to review how this PR compares to the feature implementation we are planning to launch, and we will give an update after that meeting. Thank you for your contribution.

**Comment by Ian2012** [comment]
@Daniel-hershel thanks for the information

@Colin-Fredericks yes, that's the question. I will update the implementation to match this without tests, and wait for the result of your internal discussion

**Comment by e0d** [comment]
> @Ian2012 Thanks for these changes!
> 
> Hi @jmakowski1123, could you please create a feature ticket for this PR? Or will we not need one since there is an existing issue (#1959) that is being (partially) fixed here?
> 
> CC @e0d @mphilbrick211

We shouldn't need one.  If #1959 doesn't meet the full need, we should update it.

**Comment by Daniel-hershel** [comment]
@Ian2012 I have met with our lead engineer on this project @jansenk , and my understanding is that the feature update we're about to launch will include the functionality included in this PR. The work we are finishing up will allow course authors to make a selection in studio that has 3 choices:

- align ORA due dates with subsection due date
- align ORA due dates with end of the course date
- leave ORA dates alone so they are only adjusted manually

My understanding is that this PR's functionality would essentially mirror the first option listed above. Please let me know if you think I'm missing something here and happy to keep talking it through. 

@jansenk I know you were going to look into this a little more after our chat yesterday to so feel free to amend anything I'm saying here. thanks

CC: @e0d

**Comment by Daniel-hershel** [comment]
@Ian2012 if you think there is still additional value in this PR that isn't represented in the work Jansen is doing, can you please help me understand what that additional value is?

The other question is whether anything in this PR will conflict with the release we are preparing to make. That is something that @jansenk would have to weigh in on. 

Thanks

**Comment by mariajgrimaldi** [comment]
@Daniel-hershel: if the functionality you folks have been working on includes the use case of _allow instructors to extend a student's due date based on individual due date extensions_, then I think we're good to go since it's basically what we're attempting here. Should we close this PR then? Also, let us know when you release the new version so we can test it!

**Comment by Daniel-hershel** [comment]
@jansenk can you please confirm the answer to the question @mariajgrimaldi is asking. Thanks!

**Comment by openedx-webhooks** [comment]
@Ian2012 Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-ora2&entry.867055334=https://github.com/openedx/edx-ora2/pull/2026&entry.1484655318=https://github.com/Ian2012&entry.752974735=2023-08-10+13:01&entry.1917517419=2023-08-29+14:47&entry.2133058324=No) so we can improve your experience in the future.
